clippy_lints/src/doc.rs

```diff
@@ -690,7 +690,16 @@ fn check_text(cx: &LateContext<'_>, valid_idents: &FxHashSet<String>, text: &str
         // Trim punctuation as in `some comment (see foo::bar).`
         //                                                   ^^
         // Or even as in `_foo bar_` which is emphasized.
-        let word = word.trim_matches(|c: char| !c.is_alphanumeric());
+        let word = {
+            let w = word.trim_matches(|c: char| !c.is_alphanumeric());
+            let offset = w.as_ptr() as usize - word.as_ptr() as usize;
+            // Preserve an initial `::` prefix, even as in `_::a::b_` which is emphasized.
+            if offset >= 2 && &word.as_bytes()[offset - 2..offset] == b"::" {
+                &word[offset - 2..offset + w.len()]
+            } else {
+                w
+            }
+        };
```

Comment:
Using pointers casted to `usize` seems very error prune to me, even if I know exactly what this is doing. The usage of since indexing on the string seems a bit complicated and has caused some errors in the past due to byte and string index mismatches. Could this entire block maybe be replaced by changing the `trim_matches` statement like this?

```suggestion
        let word = word.trim_matches(|c: char| !c.is_alphanumeric() && c != ':');
```

That would allow for identifiers like `:::::::way_to_many_colons::` but that would the current as well, if I'm seeing this correctly. (It's also debatable if this would be an FP)

Reply:
I wanted to avoid those too many colons. But I agree with you that this leads to convoluted code, I don't like this pointer arithmetic much myself. Maybe I should use different trims for the start and the end (this won't take more time) so that at least we don't have trailing `:`. I'll see if removing `:` as long as the resulting string starts with `:::` (so that we keep only two colons) leads to clean code.

Reply:
The trailing `:` is a good point. The code could either use `trim_start_matches` and `trim_end_matches` to only allow the prefix or validate that the word at least ends with two colons. I think the later option would make more sense. Then we can still say:

```rs
/// Import an item from ::awesome::global::blob:: (Intended postfix)
/// These are the options for ::Cat: (Indended trailing single colon, shouln't be linted)
```

Reply:
I was testing that `and:::std::mem::forget` was correctly parsed as `::std::mem::forget` but it is indeed parsed as `and:::std::mem::forget` since we don't check that we don't have more than two `::` in a row. So right now I won't remove extra `:` at the beginning, but maybe if we add a check that we don't have `:::` sequences to determine if a word is valid it would be worth filtering extra `:` at the beginning of the word.

Reply:
I also think that it's fine if such cases are linted. If someone writes and:::std::mem::forget it should be pointed out :upside_down_face:

Reply:
And as you said, this was also previously accepted, and we didn't receive any complaints about that as far as I know :upside_down_face: